torch/csrc/jit/script/compiler.cpp

```diff
@@ -559,25 +593,95 @@ struct to_ir {
     emitLoopCommon(stmt.range(), {}, {cond}, stmt.body(), {});
   }
 
+  // Validate that the `lhs` Expr's in an assignment statement are valid. That
+  // is:
+  //
+  // 1) All lhs Expr's are either Var or Starred nodes
+  // 2) There is at most one Starred node in the lhs Expr
+  // 3) A Starred node can only appear when there is another non-Starred lhs Expr
+  //    Concretely this means that `*abc = func()` is illegal. Unpacking all
+  //    outputs into a tuple is covered by `abc = func()`.
+  size_t calcNumStarredUnpack(const List<Expr>& lhs, const SourceRange& r) {
+    size_t num_normal_assign = 0;
+    size_t num_starred = 0;
+    for (const auto& assignee : lhs) {
+      if (assignee.kind() == TK_VAR) {
+        num_normal_assign++;
+      } else if (assignee.kind() == TK_STARRED) {
+        num_starred++;
+      } else {
+        throw ErrorReport(assignee)
+            << "lhs of assignment must be a variable or starred expression.";
+      }
+    }
+
+    if (num_starred > 1) {
+      throw ErrorReport(r)
+          << "Only one starred expression is allowed on the lhs.";
+    }
+
+    if (num_starred > 0 && num_normal_assign == 0) {
+      throw ErrorReport(r) << "A Starred expression may only appear on the "
+                              << "lhs within the presence of another non-starred"
+                              << " expression.";
+    }
+
+    return num_starred;
+  }
+
+  std::vector<SugaredValuePtr> createSugaredValuesFromValues(const std::vector<Value*> values) {
+    std::vector<SugaredValuePtr> sugared_outputs;
+    sugared_outputs.reserve(values.size());
+    for (Value* output : values) {
+      sugared_outputs.emplace_back(std::make_shared<SimpleValue>(output));
+    }
+    return sugared_outputs;
+  }
+
   std::vector<Value*> emitAssignment(const Assign& stmt) {
     std::vector<Value*> outputs{stmt.lhs().size()};
+    size_t num_starred_unpack = calcNumStarredUnpack(stmt.lhs(), stmt.range());
     if (stmt.reduction() != '=') {
       if (stmt.lhs().size() != 1) {
         throw ErrorReport(stmt)
-            << "reductions are only allow when there is a single variable "
+            << "reductions are only allowed when there is a single variable "
             << "on the left-hand side.";
       }
-      Ident lhs = stmt.lhs()[0];
+      Ident lhs = Var(stmt.lhs()[0]).name();
       Expr expr = BinOp::create(stmt.range(), stmt.reduction(),
                                 Var::create(lhs.range(), lhs), stmt.rhs());
-      outputs = emitExpr(expr, 1);
+      outputs = emitExpr(expr, {1, false});
     } else {
-      outputs = emitExpr(stmt.rhs(), stmt.lhs().size());
+      CallsiteDescriptor cd{stmt.lhs().size(), !!num_starred_unpack || stmt.lhs().size() == 1};
```

Comment:
`!!num_starred_unpack`? Looks like an extra reason to make the type of that return a `bool`